File: java/formula-java/src/main/java/org/formula/ResolvedListValue.java
package org.formula;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

class ResolvedListValue extends ResolvedValue {

    private final List<ResolvedValue> values;

    ResolvedListValue(List<ResolvedValue> values) {
        this.values = new ArrayList<>(values);
    }

    @Override
    public String asText() {
        return latest().asText();
    }

    @Override
    public int asNumber() {
        return latest().asNumber();
    }

    @Override
    public double asDecimal() {
        return latest().asDecimal();
    }

    @Override
    public boolean asBoolean() {
        return latest().asBoolean();
    }

    @Override
    public List<ResolvedValue> asList() {
        return Collections.unmodifiableList(values);
    }

    private ResolvedValue latest() {
        if (values.isEmpty()) {
            return ResolvedValue.none();
        }
        return values.get(values.size() - 1);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof ResolvedListValue that)) {
            return false;
        }
        return Objects.equals(values, that.values);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(values);
    }

    @Override
    public String toString() {
        return values.toString();
    }
}

File: formula-js/src/ResolvedListValue.ts
import {ResolvedValue} from "./ResolvedValue";

export class ResolvedListValue implements ResolvedValue {
  constructor(private readonly values: ResolvedValue[]) {
  }

  asText(): string {
    return this.latest().asText();
  }

  asNumber(): number {
    return this.latest().asNumber();
  }

  asBoolean(): boolean {
    return this.latest().asBoolean();
  }

  asList(): ResolvedValue[] {
    return this.values;
  }

  equals(other: ResolvedValue): boolean {
    const otherList = other.asList();
    if (otherList.length !== this.values.length) {
      return false;
    }
    for (let i = 0; i < this.values.length; i++) {
      if (!this.values[i].equals(otherList[i])) {
        return false;
      }
    }
    return true;
  }

  toString(): string {
    return this.values.toString();
  }

  map(fn: (value: ResolvedValue) => ResolvedValue): ResolvedValue {
    return fn(this);
  }

  hasValue(): boolean {
    return true;
  }

  private latest(): ResolvedValue {
    if (this.values.length == 0) {
      return ResolvedValue.None;
    }
    return this.values[this.values.length - 1];
  }
}